abmarl/sim/gridworld/actor.py

```diff
@@ -147,69 +166,366 @@ def attack_mapping(self, value):
+
+        1. The attackable agent is active.
+        2. The attackable agent is positioned at the attacked cell.
+        3. The attackable agent is valid according to the attack_mapping.
+        4. The attacking agent's accuracy is high enough.
+
+        Furthemore, a single agent may only be attacked once if stacked_attacks
+        is False. Additional attacks will be applied on other agents or wasted.
+
+        If the attack is successful, then the attacked agent's health is depleted
+        by the attacking agent's strength, possibly resulting in its death.
         """
-        return AttackingAgent
+        if isinstance(attacking_agent, self.supported_agent_type):
+            action = action_dict[self.key]
+            attacked_agents = self._determine_attack(attacking_agent, action)
+            for attacked_agent in attacked_agents:
+                if not attacked_agent.active: continue # Skip this agent since it is dead
+                attacked_agent.health = attacked_agent.health - attacking_agent.attack_strength
+                if not attacked_agent.active:
+                    self.grid.remove(attacked_agent, attacked_agent.position)
+            return attacked_agents
+        return []
 
-    def process_action(self, attacking_agent, action_dict, **kwargs):
+    def _basic_criteria(self, attacking_agent, candidate):
+        """
+        Basic criteria shared among attack actors for attack success.
+
+        The basic criteria shared amoung all attacking actors is the following:
+        1. The candidate must not be the same as the attacking agent.
+        2. The candidate must be active.
+        3. The candidate's encoding must be in the attacker's attack mapping.
+        4. Attempt the attack by rolling the dice against the attacker's accuracy.
+
+        If each of these criteria's is met, then the attack is succesful.
+
+        Args:
+            attacking_agent: The attacking agent.
+            candidate: The candidate agent.
+        Returns:
+            True if the attack is successful, otherwise False.
+        """
+        if candidate.id == attacking_agent.id: # Cannot attack yourself
+            return False
+        elif not candidate.active: # Cannot attack inactive agents
+            return False
+        elif candidate.encoding not in self.attack_mapping[attacking_agent.encoding]:
+            # Cannot attack this type of agent
+            return False
+        elif np.random.uniform() > attacking_agent.attack_accuracy:
+            # Failed attack
+            return False
+        else:
+            return True
+
+    def _subset_attackables(self, attackable_agents, number_of_attacks):
+        """
+        Subset the list of attackable agents by the number of attacks.
+
+        If the number of attacks is greater than the list of agents and stacked
+        attacks is False, then return the entire list of agents. Otherwise, randomly
```

Comment:
...then each attackable agent is attacked, and any extra attacks are wasted. Otherwise, ...